docroot/sites/emergency.uiowa.edu/modules/emergency_core/src/EmergencyAPI.php

```diff
@@ -0,0 +1,126 @@
+use Psr\Log\LoggerInterface;
+
+/**
+ * Emergency API service.
+ */
+class EmergencyAPI {
+  // Stage endpoint with test data.
+  const BASE_URL_1 = 'https://content.getrave.com/cap/uiowatest/channel1';
+  // Prod endpoint with real data.
+  const BASE_URL_2 = 'https://content.getrave.com/cap/uiowa/channel1';
+
+  /**
+   * The emergency_core logger channel.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  protected $logger;
+
+  /**
+   * The emergency_core cache.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  protected $cache;
+
+  /**
+   * The HTTP client.
+   *
+   * @var \GuzzleHttp\ClientInterface
+   */
+  protected $client;
+
+  /**
+   * Constructs an alert object.
+   *
+   * @param \Psr\Log\LoggerInterface $logger
+   *   The emergency_core logger channel.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache
+   *   The emergency_core cache.
+   * @param \GuzzleHttp\ClientInterface $http_client
+   *   The HTTP client.
+   */
+  public function __construct(LoggerInterface $logger, CacheBackendInterface $cache, ClientInterface $http_client) {
+    $this->logger = $logger;
+    $this->cache = $cache;
+    $this->client = $http_client;
+  }
+
+  /**
+   * Make an Emergency API request and return data.
+   *
+   * @param string $method
+   *   The HTTP method to use.
+   * @param array $options
+   *   Optional request options. All requests expect JSON response data.
+   * @param string $base
+   *   The base URL to use for the request. Defaults to self::BASE_URL_1.
+   *
+   * @return mixed
+   *   The API response data.
+   */
+  public function request($method, array $options = [], $base = self::BASE_URL_1) {
+
+    // Merge additional options with default but allow overriding.
+    $options = array_merge([
+      'headers' => [
+        'Accept' => 'application/xml',
+      ],
+    ], $options);
+
+    // Create a hash for the CID. Can always be decoded for debugging purposes.
+    $hash = base64_encode($base . serialize($options));
+    $cid = "emergency_core:request:{$hash}";
+    // Default $data to FALSE in case of API fetch failure.
+    $data = FALSE;
+
+    if ($cache = $this->cache->get($cid)) {
+      $data = $cache->data;
+    }
+    else {
+      try {
+        $response = $this->client->request($method, $base, $options);
+      }
+      catch (RequestException | GuzzleException $e) {
+        $this->logger->error('Error encountered getting data from @endpoint: @code @error', [
+          '@endpoint' => $base,
+          '@code' => $e->getCode(),
+          '@error' => $e->getMessage(),
+        ]);
+      }
+
+      if (isset($response)) {
+        $contents = $response->getBody()->getContents();
+        $alert = simplexml_load_string($contents);
+        $json = json_encode($alert);
+
+        /** @var object $data */
+        $data = json_decode($json, TRUE);
+
+        // Cache for 1 minute.
+        $this->cache->set($cid, $data, time() + 60);
```

Comment:
if we need to cache at all, i'd argue we want this to be less than a minute so there isn't a chance that the scheduled job doesn't call it and serve a cached version, waiting a whole minute more to get the latest data. i think the simple fix would be to change it to 30. if there is no use-case to cache though, i think we could just remove all of the caching since this isn't visitor-triggered

Reply:
Great catch! I don't think there is a use-case to cache, I was just following previous work that updates far less frequently. I will remove it.